crates/iceberg/src/metrics.rs

```diff
@@ -0,0 +1,154 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This module contains the metrics reporting API for Iceberg.
+//!
+//! It is used to report table operations in a pluggable way. See the [docs]
+//! for more details.
+//!
+//! [docs] https://iceberg.apache.org/docs/latest/metrics-reporting
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_trait::async_trait;
+use tracing::info;
+
+use crate::TableIdent;
+use crate::expr::Predicate;
+use crate::spec::SchemaId;
+
+/// This trait defines the API for reporting metrics of table operations.
+///
+/// Refer to the [Iceberg docs] for details.
+///
+/// [Iceberg docs]: https://iceberg.apache.org/docs/latest/metrics-reporting/
+#[async_trait]
+pub(crate) trait MetricsReporter: Debug + Send + Sync {
+    /// Indicates that an operation is done by reporting a MetricsReport.
+    ///
+    /// Any errors are expected to be handled internally.
+    async fn report(&self, report: MetricsReport);
+}
+
+/// An enum of all metrics reports.
+#[derive(Debug)]
+pub(crate) enum MetricsReport {
+    /// A Table Scan report that contains all relevant information from a Table Scan.
+    Scan {
+        table: TableIdent,
+        snapshot_id: i64,
+        schema_id: SchemaId,
+
+        /// If None, the scan is an unfiltered full table scan.
+        filter: Option<Arc<Predicate>>,
+
+        /// If None, the scan projects all fields.
+        // TODO: We could default to listing all field names in those cases: check what Java is doing.
+        projected_field_names: Option<Vec<String>>,
+        projected_field_ids: Arc<Vec<i32>>,
+
+        metrics: Arc<ScanMetrics>,
+        metadata: HashMap<String, String>,
+    },
+}
+
+/// Carries all metrics for a particular scan.
+#[derive(Debug)]
+pub(crate) struct ScanMetrics {
```

Comment:
Note that the Java implementation uses special types for the metrics (e.g. [`TimerResult.java`](https://github.com/apache/iceberg/blob/30ee7e83cfa3e5559d8d3ba4b6380156accb43f1/core/src/main/java/org/apache/iceberg/metrics/TimerResult.java#L28) and [`CounterResult.java`](https://github.com/apache/iceberg/blob/30ee7e83cfa3e5559d8d3ba4b6380156accb43f1/core/src/main/java/org/apache/iceberg/metrics/CounterResult.java#L27)). They include a value and a unit but I felt like both the ScanMetrics' field names and their types should convey everything we need. The RestMetricReporter will need to emit reports that follow this format but I omitted it from the general purpose ScanMetrics. 

Happy for any feedback!